homeassistant/components/nws/__init__.py

```diff
@@ -1 +1,195 @@
+    """Set up the National Weather Service integration."""
+    if DOMAIN not in config:
+        return True
+
+    hass.data[DOMAIN] = hass.data.get(DOMAIN, {})
+    for entry in config[DOMAIN]:
+        latitude = entry.get(CONF_LATITUDE, hass.config.latitude)
+        longitude = entry.get(CONF_LONGITUDE, hass.config.longitude)
+        api_key = entry[CONF_API_KEY]
+        station = entry.get(CONF_STATION)
+        scan_interval = entry.get(CONF_SCAN_INTERVAL, 10)
+
+        client_session = async_get_clientsession(hass)
+
+        if unique_id(latitude, longitude) in hass.data[DOMAIN]:
+            _LOGGER.error(
+                "Duplicate entry in config: latitude %s  latitude: %s",
+                latitude,
+                longitude,
+            )
+            continue
+
+        nws_data = NwsData(hass, latitude, longitude, api_key, client_session)
+
+        try:
+            await nws_data.async_set_station(station)
+        except (aiohttp.ClientError, asyncio.TimeoutError) as err:
+            _LOGGER.error("Error automatically setting station: %s", str(err))
+            return False
+
+        await nws_data.async_update()
+
+        hass.data[DOMAIN][unique_id(latitude, longitude)] = nws_data
+
+        scan_time = datetime.timedelta(minutes=scan_interval)
+        async_track_time_interval(hass, nws_data.async_update, scan_time)
+
+        for component in PLATFORMS:
+            hass.async_create_task(
+                discovery.async_load_platform(hass, component, DOMAIN, {}, config)
+            )
+
+    return True
+
+
+class NwsData:
+    """Data class for National Weather Service integration."""
+
+    def __init__(self, hass, latitude, longitude, api_key, websession):
+        """Initialize the data."""
+        self.hass = hass
+        self.latitude = latitude
+        self.longitude = longitude
+        ha_api_key = f"{api_key} homeassistant"
+        self.nws = SimpleNWS(latitude, longitude, ha_api_key, websession)
+
+        self.update_observation_success = True
+        self.update_forecast_success = True
+        self.update_forecast_hourly_success = True
+
+    async def async_set_station(self, station):
+        """
+        Set to desired station.
+
+        If None, nearest station is used.
+        """
+        await self.nws.set_station(station)
+        _LOGGER.debug("Nearby station list: %s", self.nws.stations)
+
+    @property
+    def station(self):
+        """Return station name."""
+        return self.nws.station
+
+    @property
+    def observation(self):
+        """Return observation."""
+        return self.nws.observation
+
+    @property
+    def forecast(self):
+        """Return day+night forecast."""
+        return self.nws.forecast
+
+    @property
+    def forecast_hourly(self):
+        """Return hourly forecast."""
+        return self.nws.forecast_hourly
+
+    async def async_update(self, now=None):
+        """Update all data."""
```

Comment:
There is a bit of code duplication here.

How about breaking this out into a function that updates each bit of data and sets the status that you could call 3 times?

Reply:
Good suggestion.